cli/cli.go

```diff
@@ -1,456 +1,822 @@
+		}); err != nil {
+			return err
+		}
+		if err := a.RunContainers(actions.RunContainersOptions{
+			GenerationPath: o.generationPath,
+			Services:       services,
+		}); err != nil {
+			return err
+		}
+		// TODO: Final tips
+		log.Error("show final tips")
 	}
-
-	return &clientImages, nil
+	return nil
 }
 
-func runCliCmd(cmd *cobra.Command, args []string, flags *CliCmdFlags, clientImages *clientImages, cmdRunner commands.CommandRunner, prompt prompts.Prompt, serviceManager services.ServiceManager, sedgeActions actions.SedgeActions) []error {
-	// Warnings
-	// Warn if custom images are used
-	if clientImages.execution != "" {
-		log.Warn(configs.CustomExecutionImagesWarning)
+func generateKeystore(p prompter.Prompter, o *CliCmdOptions, a actions.SedgeActions) error {
+	if err := selectKeystoreSource(p, o); err != nil {
+		return err
 	}
-	if clientImages.consensus != "" {
-		log.Warn(configs.CustomConsensusImagesWarning)
+	switch o.keystoreSourceType {
+	case SourceTypeSkip:
+		return nil
+	case SourceTypeCreate:
+		// Get the mnemonic
+		if err := selectKeystoreMnemonicSource(p, o); err != nil {
+			return err
+		}
+		switch o.keystoreMnemonicSource {
+		case SourceTypeCreate:
+			candidate, err := keystores.CreateMnemonic()
+			if err != nil {
+				return err
+			}
+			if err := saveMnemonic(a.GetCommandRunner(), candidate); err != nil {
+				return err
+			}
+			o.keystoreMnemonic = candidate
+		case SourceTypeExisting:
+			if err := inputKeystoreMnemonicPath(p, o); err != nil {
+				return err
+			}
+		}
+		// Get the passphrase
+		if err := selectKeystorePassphraseSource(p, o); err != nil {
+			return err
+		}
+		switch o.keystorePassphraseSource {
+		case SourceTypeExisting:
+			if err := inputKeystorePassphrasePath(p, o); err != nil {
+				return err
+			}
+			if passphrase, err := readFileContent(o.keystorePassphrasePath); err != nil {
+				return err
+			} else {
+				o.keystorePassphrase = passphrase
+			}
+		case SourceTypeCreate:
+			if err := inputKeystorePassphrase(p, o); err != nil {
+				return err
+			}
+		}
+		if err := runPromptActions(p, o,
+			inputWithdrawalAddress,
+			inputNumberOfValidators,
+			inputNumberOfExistingValidators,
+		); err != nil {
+			return err
+		}
+		if err := keystores.CreateKeystores(keystores.ValidatorKeysGenData{
+			Mnemonic:    o.keystoreMnemonic,
+			Passphrase:  o.keystorePassphrase,
+			OutputPath:  filepath.Join(o.generationPath, "keystores"),
+			MinIndex:    uint64(o.existingValidators),
+			MaxIndex:    uint64(o.existingValidators) + uint64(o.numberOfValidators),
+			NetworkName: o.genData.Network,
+			ForkVersion: configs.NetworksConfigs()[o.genData.Network].GenesisForkVersion,
+			// Constants
+			UseUniquePassphrase: true,
+			Insecure:            false,
+			AmountGwei:          uint64(eth2.Mainnet.MAX_EFFECTIVE_BALANCE),
+			AsJsonList:          true,
+		}); err != nil {
+			return err
+		}
+
+	case SourceTypeExisting:
+		if err := inputKeystorePath(p, o); err != nil {
```

Comment:
Same. This check is heavier than the others bc of the keystone structure

Reply:
It seems that no validation over the name of the `validator_key` folder is happening. Also, the code is sending an error up to the runtime stack, and I see this error will be log.Fatal()-led instead of a graceful exit.

Reply:
Variable [`ValidatorKeysDirName`](https://github.com/NethermindEth/sedge/blob/4187889fbab7076e9c51f66d5aeef82bc31d308a/internal/pkg/keystores/validate.go#L46) contains the expected name for the folder containing the validator keys files. This folder is checked [here](https://github.com/NethermindEth/sedge/blob/4187889fbab7076e9c51f66d5aeef82bc31d308a/internal/pkg/keystores/validate.go#L80-L84), if not found, then an `ErrValidatorKeysNotFound` is returned. This check covers the case where the folder has a different name.

Reply:
The `cli` command returns the error, and the fatal log happens in the root command [here](https://github.com/NethermindEth/sedge/blob/4187889fbab7076e9c51f66d5aeef82bc31d308a/cmd/sedge/main.go#L78). Errors coming from any command should be handled there, maybe with a simple error log and then return with exit code 1 instead of fatal

Reply:
Ok, all good. I changed my mind about the error thing. `log.Fatal` is a `log.Error()` followed by a `os.Exit(1)`